torch/csrc/jit/codegen/cuda/ir_nodes.cpp

```diff
@@ -1031,6 +1030,279 @@ std::pair<TensorDomain*, TensorDomain*> TensorDomain::rFactor(
+  int fixedPoint(int e) {
+    TORCH_INTERNAL_ASSERT(SetMap.size() > e);
+    while (SetMap[e] != e)
+      e = SetMap[e] = SetMap[SetMap[e]];
+    return e;
+  }
+
+  // utility to check the class I belongs to:
+  //  returns the equivalent class that e belongs to
+  //  will create a new class if no match seen
+  int fixedPoint(T I) {
+    if (!EntryMap.count(I))
+      createPoint(I);
+    return fixedPoint(EntryMap[I]);
+  }
+
+  // utility to create a new equiv class for I
+  void createPoint(T I) {
+    EntryMap[I] = count;
+    SetMap.push_back(count++);
+    Weights.push_back(1);
+  }
+}; // class DisjoinSet
+
+/** \brief Concretize broadcast axes, i.e. identifying a non-broadcast
+ * IterDomain that the broadcast IterDomain can map to.
+ *  \details This traversal
+ * processes root domains only, concretization works by inspecting pointwise
+ * ops, e.g. : T2 [i0,i1] = T1[i0,B0] + T0[i0,i1] will concretize axis B0 to i1
+ */
+class ConcretizeDomain : public BackwardVisitor {
+ public:
+  /** \brief traverses the graph backward from outputs
+   *         to identify all concretizing opportunities
+   */
+  ConcretizeDomain(Fusion* fusion) {
+    traverseFrom(fusion, fusion->outputs(), true);
+  }
+
+  /** \brief API call to run the concretize pass and return the
+   *        axis that bcastDom concretizes to
+   */
+  static const IterDomain* getConcreteDomain(IterDomain* bcastDom) {
+    ConcretizeDomain CD(bcastDom->fusion());
+
+    // remove this assertion once we support broadcast on output
+    TORCH_INTERNAL_ASSERT(CD.canConcretize(bcastDom));
+    return CD.concretized(bcastDom);
+  }
+
+  // returns true if either id is not a broadcast or
+  // the traversal has found a concretized axis for id
+  bool canConcretize(IterDomain* id) {
+    return !id->isBroadcast() || BcastDomainMap_.count(id);
+  }
+
+  // returns the concretized id recorded from traversal
+  IterDomain* concretized(IterDomain* id) {
+    TORCH_INTERNAL_ASSERT(canConcretize(id));
+    if (!id->isBroadcast())
+      return id;
+    return BcastDomainMap_.at(id);
+  }
+
+ protected:
+  using MapType = std::unordered_map<IterDomain*, IterDomain*>;
+  MapType BcastDomainMap_;
+
+  // utility to inspect a pointwise operator and
+  // record concretize opportunities
+  void concretizePwOp(Expr* e);
+
+  // utility to record new concretize opportunity
+  void concretizeTo(IterDomain* id, IterDomain* To) {
+    TORCH_INTERNAL_ASSERT(id->isBroadcast() && !To->isBroadcast());
+    BcastDomainMap_[id] = concretized(To);
+  }
+
+  void handle(UnaryOp* uop) override {
+    concretizePwOp(uop->asExpr());
+  }
+
+  void handle(BinaryOp* bop) override {
+    concretizePwOp(bop->asExpr());
+  }
+
+  void handle(TernaryOp* top) override {
+    concretizePwOp(top->asExpr());
+  };
+}; // class ConcretizeDomain
+
+void ConcretizeDomain::concretizePwOp(Expr* e) {
+  TensorView* TVo = *ir_utils::filterByType<TensorView>(e->outputs()).begin();
```

Comment:
Just `tv`? I believe we want variable names to start with small caps.

Reply:
Thanks for the suggestion. Updated the variable naming.